lib/iex/lib/iex/broker.ex

```diff
@@ -0,0 +1,181 @@
+defmodule IEx.Broker do
+  @moduledoc false
+  @name IEx.Broker
+
+  @type take_ref :: {takeover_ref :: reference(), server_ref :: reference()}
+
+  use GenServer
+
+  ## Shell API
+
+  @doc """
+  Finds the IEx server running inside `:user_drv`, on this node exclusively.
+  """
+  @spec shell :: pid | nil
+  def shell() do
+    # Locate top group leader, always registered as user
+    # can be implemented by group (normally) or user
+    # (if oldshell or noshell)
+    if user = Process.whereis(:user) do
+      case :group.interfaces(user) do
+        # Old or no shell
+        [] ->
+          case :user.interfaces(user) do
+            [] -> nil
+            [shell: shell] -> shell
+          end
+
+        # Get current group from user_drv
+        [user_drv: user_drv] ->
+          case :user_drv.interfaces(user_drv) do
+            [] -> nil
+            [current_group: group] -> :group.interfaces(group)[:shell]
+          end
+      end
+    end
+  end
+
+  @doc """
+  Finds the evaluator and server running inside `:user_drv`, on this node exclusively.
+  """
+  @spec evaluator :: {evaluator :: pid, server :: pid} | nil
+  def evaluator() do
+    if pid = shell() do
+      {:dictionary, dictionary} = Process.info(pid, :dictionary)
+      {dictionary[:evaluator], pid}
+    end
+  end
+
+  ## Broker API
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, :ok, name: @name)
+  end
+
+  @doc """
+  Register an IEx server in the broker.
+
+  All instances, except shell ones, are registered.
+  """
+  @spec register(pid) :: :ok
+  def register(pid) do
+    GenServer.call(@name, {:register, pid})
+  end
+
+  @doc """
+  Respond to a broker take over request.
```

Comment:
```suggestion
  Client responds to a takeover request.
```

I have doubts regarding the usage of "takeover" vs "take over" on this one